apps/src/code-studio/components/header/SignInCallout.jsx

```diff
@@ -0,0 +1,135 @@
+import React from 'react';
+import cookies from 'js-cookie';
+
+const CALLOUT_COLOR = '#454545';
+const TRIANGLE_BASE = 30;
+const TRIANGLE_HEIGHT = 15;
+const CALLOUT_Z_INDEX = 1040;
+const CALLOUT_TOP = 30;
+
+const HideSignInCallout = 'hide_signin_callout';
+
+const styles = {
+  container: {
+    // The outermost div is relatively positioned so it can be used as a positional
+    // anchor for its children (which will be absolutely positioned to avoid affecting
+    // layout).  This element must be 0-sized to avoid affecting layout.
+    position: 'relative',
+    height: 0,
+    width: 0
+  },
+  content: {
+    position: 'absolute',
+    top: CALLOUT_TOP,
+    right: -90,
+    zIndex: CALLOUT_Z_INDEX,
+    backgroundColor: CALLOUT_COLOR,
+    borderRadius: 3
+  },
+  modalBackdrop: {
+    // Most backdrop attributes come from the 'modal-backdrop' class defined by bootstrap
+    // but we need to override the opacity as the default opacity of 0.8 is too dark.
+    // Note that bootstrap defaults the z-index of the backdrop to 1040.
+    opacity: 0.5
+  },
+  upTriangle: {
+    position: 'absolute',
+    top: CALLOUT_TOP - TRIANGLE_HEIGHT,
+    left: -(TRIANGLE_HEIGHT / 2.0),
+    width: 0,
+    height: 0,
+    borderStyle: 'solid',
+    borderTopWidth: 0,
+    borderRightWidth: TRIANGLE_BASE,
+    borderBottomWidth: TRIANGLE_HEIGHT,
+    borderLeftWidth: TRIANGLE_BASE,
+    borderTopColor: 'transparent',
+    borderRightColor: 'transparent',
+    borderBottomColor: CALLOUT_COLOR,
+    borderLeftColor: 'transparent',
+    zIndex: CALLOUT_Z_INDEX
+  },
+  contentContainer: {
+    display: 'flex',
+    flexDirection: 'row',
+    padding: 20
+  },
+  imageContainer: {
+    width: 100,
+    marginRight: 20
+  },
+  textContainer: {
+    width: 400,
+    textAlign: 'left',
+    whiteSpace: 'normal'
+  },
+  textHeader: {
+    marginTop: 0
+  }
+};
+
+/*
+ * This is a callout attached to the sign-in button that's used on CSF level
+ * pages to remind the user to sign-in.  Note that the sign-in button is
+ * defined in shared/haml/user_header.haml and is not a React component.
+ * This component is injected into the page by src/code-studio/header.js.
+ */
+export default class SignInCallout extends React.Component {
+  constructor(props) {
+    super(props);
+
+    this.closeCallout = this.closeCallout.bind(this);
+    this.getContent = this.getContent.bind(this);
+
+    // Using explicit strings: only show if neither says to hide
+    this.state = {
```

Comment:
It feels a bit weird to me that the cookie logic is mixed in with the UI logic in this component.  Ideally, I think we want a state-less callout component and then another component that acts as the glue between the callout and the sign-in button (and would be where the cookie logic lives).  At the same time, I don't want to over-complicate this.  Curious to know if @cforkish and @maureensturgeon have thoughts on this?

Reply:
I see what you mean. And as you mentioned, I like the readability/simplicity of having it here, but understand if it would be better practice to separate out. Will pull those two in to review as well^

Reply:
Yes, I agree I would expect a SignInCallout component not to handle the logic of whether it's shown. I would expect a parent component to have logic to determine whether it's displayed and that to render the component, and the SignInCallout component would just render the callout.

Reply:
@maureensturgeon @jamescodeorg Would it be bad practice to merge this and resolve in the next PR (which is meant for handling the component's connection to the actual header)? If not, can rework now, and will simply pull the logic out of here completely.

Reply:
my initial thought is that the cookie/storage logic should live in redux, and the parent component should subscribe to a `showSignInCallout` bool in the store. if @jamescodeorg and @maureensturgeon think that's worth exploring i'd be happy to think through it with you! but either way, i agree that this logic should be in the parent.

fwiw, in this type of UI architecture, a common pattern is to do what @jamescodeorg describes with redux -- so you have a stateless UI component wrapped in a "business logic component" that connects to redux. [this is what that looks like in the Flutter framework](https://codete.com/blog/building-with-blocs-sharing-the-code/), but in our case you can just substitute the "Sink/Stream" stuff with redux connect.

Reply:
But yeah, in terms of the other cookie logic, I could see us having that in a separate file that has maybe a `getHideSignInCalloutCookie` and `setHideSignInCalloutCookie` method which are exported?

Reply:
i agree that doing this in redux might be overhead overkill, but i also believe that a little extra overhead is often worth it in the name of adopting reusable patterns. but since we don't yet have any great canonical patterns around this type of thing, i'm not going to push for doing this in redux.

sans redux, the two options for separating the business logic from the ui are a separate js module (e.g. `cookieHelpers`) as Maureen suggests, and/or a wrapper component that handles the business logic itself per Jame's suggestion. 

personally, if we put the logic in a helper module, adding a parent component to read/write that module seems a little excessive, so with that approach i'd actually vote for just calling into that module from here. i don't have a problem with components that decide themselves whether or not to render when the logic is so tightly coupled to the specific component.

that said, adding a generic `<SessionDataProvider>` component to handle this sort of thing sounds like an interesting reusable option to consider -- pass in an enum for the cookie you want that provides its child with a `SessionData` object and a `setSessionData` callback, and calls into a `sessionDataHelpers` module to handle the logic for reading/writing `SessionData` objects for each enum value, e.g.
```
<SessionDataProvider sessionData={SessionData.showSignInCallout} >
  <SignInCallout />
<SessionDataProvider />
```
that sounds interesting to me from an engineering perspective, but tbph just moving the logic into a helper module to call into from this component sounds like a simple and perfectly reasonable solution to me.

Reply:
okay, feel free to ignore more at this point but i have more thoughts to share.

we use redux as _the_ way to manage non-transient state in our application. redux is designed to be the layer between your ui and any external services your application uses. in our case, that primarily just means our backend. but any data-providing third party library would also constitute such a service. even though cookies are globally accessible, to the extent that we want to use them to control application state, i strongly believe redux should still be the only state manager that our components ever reference directly. i really don't like the idea of having a hybrid codebase where some state is managed by redux and other state by plain js modules. we've adopted a specific architecture (redux), and i firmly believe it's worth a little overhead to stick to it.

thinking forward to the likely possibility that we have some `sessionStorageHelper` or `cookieHelper` module that we expand to handle additional similar use cases as needed here (with or without a `<Provider>` component, since we're specifically talking about non-transient state, that helper module would become a de-facto data store, which doesn't sit well with me.

we have a ton of work to do in terms establishing better and more reusable design patterns around our use of redux and we are preparing to develop a "north star" in that regard, but in the meantime, i still think it's worth building new functionality in a way that moves us toward -- rather than away from -- that north star and helps us figure out all the necessary use cases it will need to accommodate.

Reply:
** steps down for soapbox, abstains for any opinion on resolution to Jame's original comment ** 😳

Reply:
@jamescodeorg @cforkish @maureensturgeon To come back around to this comment party: I've separated out the logic into its own file, so the cookies and session storage will live a level above the component. The onclick call is passed down as a prop, and the component just renders itself. The testing I've added for now only test the actual look of the component, and the next PR will test the cookies, session storage, and the onclick (at minimum).